stan/math/prim/scal/fun/get_num_threads.hpp

```diff
@@ -0,0 +1,76 @@
+#ifndef STAN_MATH_PRIM_SCAL_FUN_GET_NUM_THREADS_HPP
+#define STAN_MATH_PRIM_SCAL_FUN_GET_NUM_THREADS_HPP
+
+#include <stan/math/prim/scal/err/invalid_argument.hpp>
+#include <boost/lexical_cast.hpp>
+
+#include <cstdlib>
+#include <thread>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Get number of threads to use for num_jobs jobs. The function uses
+ * the environment variable STAN_NUM_THREADS and follows these
+ * conventions:
+ *
+ * - STAN_NUM_THREADS is not defined => num_threads=1
+ * - STAN_NUM_THREADS is positive => num_threads is set to the
+ *   specified number
+ * - STAN_NUM_THREADS is set to -1 => num_threads is the number of
+ *   available cores on the machine
+ * - STAN_NUM_THREADS < -1, STAN_NUM_THREADS = 0 or STAN_NUM_THREADS is
+ *   not numeric => throws an exception
+ *
+ * Should num_threads exceed the number of jobs, then num_threads will
+ * be set equal to the number of jobs.
+ *
+ * @param num_jobs number of jobs
+ * @return number of threads to use
+ * @throws std::runtime_error if the value of STAN_NUM_THREADS env. variable
+ * is invalid
+ */
+// TODO(SW):
+// 1. get rid of num_jobs argument which is not needed to find
+// out about number of threads we need to use.
+// 2. Consider to switch from std::thread::hardware_concurrency() to
+// tbb::task_scheduler_init::default_num_threads() once TBB is
+// mandatory.
+// 3. pull out of internal?
+inline int get_num_threads(int num_jobs) {
+  int num_threads = 1;
+#ifdef STAN_THREADS
+  const char* env_stan_num_threads = std::getenv("STAN_NUM_THREADS");
```

Comment:
Sidenote: We should start thinking about how to add the number of threads and the OpenCL devices so they are flags at runtime

Reply:
Maybe... not sure. There are pros and cons for both ways (runtime environment vs interface configuration). I am continuing here the system we started for `map_rect`.

Reply:
(moving all my comments up to here so that it's easier for others to read the code from the github page)

Reply:
line 50: personally, especially with multiple if elses, I like having brackets to show leaving and entering a scope

Reply:
From [the docs for hardware_concurrency](https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency)

> Number of concurrent threads supported. If the value is not well defined or not computable, returns ​0​

Should we have a check here and if the returned value is zero return 1?

Also does this return the number of logical or physical cores? I think previously we saw that hyperthreading etc was not that useful

Reply:
The `get_num_jobs` function is right now used inside `map_rect`, but its use will go away there once we switch over `map_rect` to use the TBB.

When we switch over `map_rect` to the TBB, we can then include this function inside the `init_threadpool_tbb.hpp` file, sure. Then we should also remove the `num_jobs` argument which is really not needed. The `get_num_threads` should only parse the `STAN_NUM_THREADS` and that's it.

The value `-1` means "use all cores". I would change this to then return the TBB default for that. As I recall that is without hyper-threading as I saw that last time.